cli/internal/globwatcher/globwatcher.go

```diff
@@ -135,24 +146,55 @@ func (g *GlobWatcher) OnFileWatchEvent(ev filewatcher.Event) {
 			g.logger.Error(fmt.Sprintf("failed to check path %v against glob %v: %v", repoRelativePath, glob, err))
 			continue
 		}
-		// If this glob matches, we know that it has changed for every hash that included this glob.
+		// If this glob matches, we know that it has changed for every hash that included this glob
+		// and is not excluded by a hash's exclusion globs.
 		// So, we can delete this glob from every hash tracking it as well as stop watching this glob.
 		// To stop watching, we unref each of the directories corresponding to this glob.
 		if matches {
-			delete(g.globStatus, glob)
+			// We need to check if all the hashes exclude this path. In which case we shouldn't remove it
+			// from globStatus
+			allExcluded := true
+
 			for hashUntyped := range hashStatus {
 				hash := hashUntyped.(string)
 				hashGlobs, ok := g.hashGlobs[hash]
+				isExcluded := false
+
+				// Check if we've excluded this path by going through exclusion globs
+				for exclusionGlob := range hashGlobs.Exclusions {
+					matches, err := doublestar.Match(exclusionGlob.(string), filepath.ToSlash(repoRelativePath))
+					if err != nil {
+						g.logger.Error(fmt.Sprintf("failed to check path %v against glob %v: %v", repoRelativePath, glob, err))
+						continue
+					}
+
+					if matches {
+						isExcluded = true
+						break
+					} else {
+						allExcluded = false
+					}
+				}
+
+				// If we have excluded this path, then we skip it
+				if isExcluded {
+					continue
+				}
+
 				if !ok {
 					g.logger.Warn(fmt.Sprintf("failed to find hash %v referenced from glob %v", hash, glob))
 					continue
 				}
-				hashGlobs.Delete(glob)
+				hashGlobs.Inclusions.Delete(glob)
 				// If we've deleted the last glob for a hash, delete the whole hash entry
-				if hashGlobs.Len() == 0 {
+				if hashGlobs.Inclusions.Len() == 0 {
 					delete(g.hashGlobs, hash)
 				}
 			}
+
+			if !allExcluded {
```

Comment:
I would add a comment here. Something like "no hash excluded this change, so we can stop tracking this glob for all of them."

It might also be a good idea to rephrase this condition as a positive. Something like `if invalidatedGlobForAllHashses`, since this is at least moderately confusing logic.